include/zephyr/drivers/stepper.h

```diff
@@ -0,0 +1,313 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Carl Zeiss Meditec AG
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_STEPPER_H
+#define ZEPHYR_INCLUDE_STEPPER_H
+
+/** \addtogroup stepper
+ *  @{
+ */
+
+/**
+ * @file
+ * @brief Public API for Stepper Motor Controller
+ *
+ */
+
+#include <zephyr/kernel.h>
+#include <errno.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Stepper Motor micro step resolution options
+ */
+enum micro_step_resolution {
+	STEPPER_FULL_STEP,
+	STEPPER_MICRO_STEP_2,
+	STEPPER_MICRO_STEP_4,
+	STEPPER_MICRO_STEP_8,
+	STEPPER_MICRO_STEP_16,
+	STEPPER_MICRO_STEP_32,
+	STEPPER_MICRO_STEP_64,
+	STEPPER_MICRO_STEP_128,
+	STEPPER_MICRO_STEP_256,
+};
+
+enum stepper_direction {
+	STEPPER_DIRECTION_POSITIVE,
+	STEPPER_DIRECTION_NEGATIVE,
+};
+
+/**
+ * @typedef stepper_enable_t
+ * @brief enable or disable the stepper motor controller.
+ */
+typedef int32_t (*stepper_enable_t)(const struct device *dev, bool enable);
+
+/**
+ * @typedef stepper_move_t
+ * @brief Move the stepper motor by a given number of steps.
+ */
+typedef int32_t (*stepper_move_t)(const struct device *dev, int32_t steps);
+
+/**
+ * @typedef stepper_set_max_velocity_t
+ * @brief Set the max velocity in steps per seconds.
+ */
+typedef int32_t (*stepper_set_max_velocity_t)(const struct device *dev, uint32_t steps_per_second);
```

Comment:
As an honest question, is the minimum speed achievable by this (1 step/sec) a desired effect? 

Asking this because I can think of some applications that would require very slow speeds, some slower than 1 step/sec that would gladly accept this API, but would need some workarounds to do so. Is this a controller limitation? If not, I would suggest choosing a smaller time unit like milliseconds. 

Just as a sidenote, I think the `uint32_t` int will be greatly underused, _i.e:_ 4294967295 steps/sec would mean ~0.2ns of period, mostly unachievable in real word standards. I assume the `steps` in here are full steps as well. Moving this to step/ms would enable better fitted interval use as well, since it would go for ~4294967 steps/sec and a period of ~0.2us instead.

Reply:
If you are trying to move slower than 1 fullstep/second, then you are almost certainly using microstepping to accomplish that in order to achieve smooth motion. Generally "steps" refers to however the controller is counting steps. It isn't always full steps, but also the MCU is often blind to the step-to-microstep conversion (it usually doesn't really need to know).

So if set to 1/256 microstepping, you have many more steps per rotation. A 200 step-per-revolution motor (common 1.8 degrees per step) becomes a 51200 step-per-revolution motor. The velocity becomes scaled as well--so this is really microsteps per second. Most controllers I have encountered use seconds as the divisor (e.x. [TMC5240 table 19 pg. 50](https://www.analog.com/media/en/technical-documentation/data-sheets/TMC5240_datasheet_Rev_0.pdf#page=50)) since the resolution is scaled with microstepping.

Most controllers I have encountered also use velocity ranges of at least 16-bit but less than 32-bit (ex. [TMC5240 again 22-bit velocity pg. 119](https://www.analog.com/media/en/technical-documentation/data-sheets/TMC5240_datasheet_Rev_0.pdf#page=119)). When the velocity is actually microsteps per second, much higher numbers become possible and necessary.

Reply:
I suppose the documentation should be updated to say that the "step" unit is not really full steps but rather steps of the current microstepping setting. Otherwise, microstepping is useless if you can only move in fullsteps via the API.

Reply:
I was suspecting that `steps` in here refer to the smallest step possible (including microsteps), but either way, applications like sun followers can have really slow speeds, e.g.: I'm planning on using this driver for a GoTo telescope mount, and some speeds are hardly dependent on motor gear ratios, to simplify some of the hardware, I usually go for speeds close to 1 (full) step per 3 seconds, and get the juicy smoothness from the microstepping as a feature of the movement, not a need itself. Although I'm using a really cheap controller (A4988) I think we should be able to support such cases as well, so I guess it would suffice to treat `steps` here as full steps, and adjust for the time resolution.

Reply:
> I suppose the documentation should be updated to say that the "step" unit is not really full steps but rather steps of the current microstepping setting. Otherwise, microstepping is useless if you can only move in fullsteps via the API.

Yep, that would work as well, to me, it was too implicit that the step was a current step by the given microstep resolution and not the full step itself.

Reply:
So, @jilaypandya could you update the documentation to reflect that `steps` here refer to the current microstepping resolution and not the full step?

Reply:
I think there may be a few other places too where it explicitly says "full steps" (or doesn't say at all) but it should say "current microstep resolution." In particular `stepper_set_actual_position`, `stepper_get_actual_position`, `stepper_set_target_position`, and `stepper_enable_constant_velocity_mode`.

I can't think of a particular controller (either step/dir interface, SPI, or UART) that accepts full steps as a unit when in a microstep mode. Running in full steps removes the added resolution of the microstepping.

@wkhadgar In your particular case, if you are using the 1/16 stepping on the A4988 then you can achieve 1 full-step/3 seconds with about 5 16th-steps/second (so set the velocity to a value of 5). I think that below a velocity of 1 microstep/second it is reasonable to use the relative move or absolute target positions commands on a much lower frequency timer. You also may want to step at a time precisely synced with a clock (world clock, not MCU clock) in which case the velocity API is not the right fit anyway.

Reply:
okay the velocity from here on shall be in microsteps_per_second, i will update it tomorrow.

Reply:
@jilaypandya see my suggested changes below. I think "steps" is the best we can do since it _could_ be full steps or microsteps. If in mode `STEPPER_FULL_STEP` then steps are full steps but if in mode `STEPPER_MICRO_STEP_16` then steps are microsteps. We just need to be explicit in the documentation that "steps" is dependent on the microstep resolution.

Reply:
> I think there may be a few other places too where it explicitly says "full steps" (or doesn't say at all) but it should say "current microstep resolution." In particular `stepper_set_actual_position`, `stepper_get_actual_position`, `stepper_set_target_position`, and `stepper_enable_constant_velocity_mode`.

That's why I thought `steps` were being pictured as full steps!

>  I think that below a velocity of 1 microstep/second it is reasonable to use the relative move or absolute target positions commands on a much lower frequency timer.

Definitively! The only problem was that we had a dubious reference to steps, and as you said, in some places it referred to full steps explicitly. Having it as the microstep (explicitly) would solve this ambiguity and allow for slower than 1 step/sec movements.

Reply:
> @jilaypandya see my suggested changes below. I think "steps" is the best we can do since it _could_ be full steps or microsteps. If in mode `STEPPER_FULL_STEP` then steps are full steps but if in mode `STEPPER_MICRO_STEP_16` then steps are microsteps. We just need to be explicit in the documentation that "steps" is dependent on the microstep resolution.

Yeah that's the whole point a step is a step which corresponds to let's say 1.8°. Now a microstep is something that is dependent on the user setting. It could vary from 1.8° to 1.8°/256(Some microstep resolution). Hence setting the speed in microsteps makes much more sense. Also if you look into the datasheets normally, the velocity is in microsteps.